### DISCUSSION TITLE
Release Strategy Discussion

### DISCUSSION THREAD

**Discussion by damienjburks** [finos/common-cloud-controls#268] [Ideas]
This is a discussion post created for tracking the ideas flowing for defining a release strategy for the CCC project.

**Comment by damienjburks** [comment]
From the email chain from @sshiells-scottlogic 

I think it is worth picking up the discussion on releases / release scheduling before the next delivery WG call?
 
What does a release look like? Do we package anything up, or is it simply the controls and threats files produced by the security WG? If it is the latter then I support Eddie's suggestion of having milestones along the lines of "Virtual Machines 1.0" etc. Then we could use "Event" milestones to group services that we are targetting to complete for specifc events e.g. OSFF NY? We would then just use labels to identify each WG's contribution towards each release (as milestones are probably too much).

We also need to decide on what versioning looks like and how we control newer/updated versions.
 
Would it be worth using the Community Structure WG meeting this week to continue this discussion?

**Comment by eddie-knight** [comment]
ref: https://github.com/finos/common-cloud-controls/issues/259#issuecomment-2240573641

**Comment by damienjburks** [comment]
@sshiells-scottlogic - my thoughts on this:

A release should be defined as a grouping of artifacts that will be communicated to the public. I should define any new controls, threats, and taxonomy catalogs that have been created or updated _(after a PR is merged into main)_. What we can do is create a release that includes a high-level description of the changes for each catalog, the names of the people who contributed to it based on commit ID, and also we could include the company logos that have contributed to the cause.

The idea I have in mind for versioning includes the entire project and follows SemVer standards. Here is an example flow diagram I came up with to help you all visualize what I am thinking when we onboard new services to a project:

![flow_diagram drawio](https://github.com/user-attachments/assets/04537b52-8ac4-4929-9553-741751e8ffb7)

And for any updates that are created, we'd increment the second number. For example, if the security WG adds in new controls/threats to an existing catalog, then we'd move from 1.0.0 to 1.1.0. The same is said for the Taxonomy WG. This could also include, but is not limited to, changes to policy and guidelines, etc.

For minor changes, like grammatical errors or formatting issues, we'd increment the last number. So if there was a misspelling of a word that was caught, the release number would go from 1.0.0 to 1.0.1. Thoughts?

I'll come back to the milestones later since I believe it'll go hand and hand with version control for our artifacts/content. I think we should try and figure that out first and then discuss milestones.

**Reply by sshiells-scottlogic** [reply]
What do you have in mind for the artifacts? 

What is not clear to me is exactly what we are deliverying to people. Are we packaging files up and shipping them via some method, or are people simply going to come to the GitHub repo and take the files from there?

**Reply by eddie-knight** [reply]
I'd like to hear yalls response to two ideas that I'm partial to:

1. Separate versioning for each service artifacts
  - For example, the Taxonomy, Controls, and Tests for _Object Storage_
  - Increases delivery automation overhead while segmenting development lifecycle logistics into smaller chunks
  - Allows for more granular attribution
2. CalVer instead of SemVer
  - Even if we use a simplified CalVer, users can know how old a control set is at a glance 
  - (ie, Object Storage 23.07 was defined in July 2023)

**Comment by sshiells-scottlogic** [comment]
I agree with both points.

1) Each service is independent. Any grouping would be arbitrary and likely delay the release of services.

2) I  think as the project grows the spread of versions will be quite wide. It's not too had to see a scenario where we have version 3.6 of one control, and 1.0 of another. I can imagine people might try to sync versions (I've got 3.6 of service x, I must have 3.6 of service y).

I think incorporating the date also makes it easier to see how stale a specific service is.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by damienjburks** [reply]
@eddie-knight I understand your points. I do like the idea of versioning artifacts separately and treating each service independently to increase velocity and the rate at which we deliver. I don't foresee any issues with introducing automation with this approach. 

I'm skeptical about using CalVer for versioning the artifacts. Let's say we make 3 changes in July to a catalog for Taxonomy for Object Storage. Versioning it as Object Storage 23.07 would be confusing, especially for tracking purposes. I think it'd be best to specify the dates that they were modified.

Also, @sshiells-scottlogic I am not sure how we'd be delivering artifacts to the public. I assumed that we'd just be sharing the links to the repository. @eddie-knight  is there a plan to host the content with GitHub pages or something?

**Reply by eddie-knight** [reply]
No plans yet. I suspect we want the controls catalogue to be available via PDF.

**Reply by damienjburks** [reply]
Is there a particular reason(s) why we'd want them to be available via PDF over self-hosting it with Hugo or something that renders the Markdown files easily?

**Reply by eddie-knight** [reply]
PDFs come to mind as that is the approach used by CIS Benchmarks.

Just brainstorming. Happy to leave the decision in your hands.

**Comment by sshiells-scottlogic** [comment]
I get your point around the 3 versions in one month using that format. However, given the pace of the project I'd be surprised if we'd run into that situation.

However, if we do could append an incremental version number to second and later releaes e.g.

23.7-2




Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by eddie-knight** [reply]
Just for the sake of discussion... CalVer comes in all the flavors of the rainbow. If we anticipate multiple releases per month (which I think would be a bad idea, but that's another discussion) we could choose YY.MM.DD in that case.

**Comment by sshiells-scottlogic** [comment]
I feel like multiple releases per month will be the exception, not the rule. I would be wary adding the .DD part as standard as I think it will just add noise to 99% of the versions that isn't required.

**Reply by damienjburks** [reply]
I think this idea.

**Comment by robmoffat** [comment]
SemVer is really all about representing backwards compatibility.  Is backwards compatibility something we care about for CCC?

**Reply by damienjburks** [reply]
I agree with this statement. We don't need to worry about backward compatibility for CCC artifacts; I'm sold on the CalVer idea. I believe it's fits the needs for this project.

**Comment by sshiells-scottlogic** [comment]
Another advatnave of CalVer is that it is easy to see at a glance how recent/old a version is. E.g. if a vuneralbility is discovered in VMs in December 24 and the VM controls were versions as 24.8 it would be clear that controls wouldn't cover that vunerability.

**Comment by damienjburks** [comment]
Okay. So we're all in agreement that:
1. We should use the CalVer standard for version control e.g Object Storage 24.07
2. Versioning is scoped to each artifact delivered by the working groups.
3. Releases should happen (at most) one time a month.

If so, then we can discuss the delivery of the artifacts. Let me know if I'm missing anything.

**Reply by eddie-knight** [reply]
Thanks for this summary 💥 

Last week the group was chatting on one of the calls about the contents of a release. It was suggested that a release should contain the service description, control description, and feature files. Does that match with these notes, or is there anything more to discuss on that front?

**Reply by damienjburks** [reply]
@eddie-knight could you drop the mtg issue number here so I can review the meeting notes if applicable? Not sure which call it was.

**Reply by eddie-knight** [reply]
Doesn't look like we took any notes on any of last week's calls... No bueno.

**Reply by damienjburks** [reply]
All good! I believe we should do a soft release with the service and control descriptions and their corresponding feature files. I'm assuming this will include the deliverables from the Taxonomy and Security WG's only.

**Reply by eddie-knight** [reply]
Sounds right to me! Do we need to capture any of this in documentation?

And are ready to mark this as the answer and close the discussion?

**Reply by damienjburks** [reply]
I am fine with marking this discussion as closed and starting the documentation process. @robmoffat @sshiells-scottlogic what do you guys think?

**Reply by damienjburks** [reply]
Bumping this up - y'all still okay with me closing this out? If so, I'll start writing the documentation to start with the versioning criteria for the project.

**Reply by eddie-knight** [reply]
Sure

**Reply by damienjburks** [reply]
awesome! i'll get to work on documenting this.
